### ISSUE TITLE
vpc.fromExistingIds() does not work with Resource Access Manager Shared VPCs/Subnets

### ISSUE THREAD

**Issue by binhrobles** [pulumi/pulumi-awsx#598]
# Description
I'm working within an organization that just began using AWS Organizations and Resource Access Manager to create/vend [shared network resources](https://docs.aws.amazon.com/vpc/latest/userguide/vpc-sharing.html). Before this, network configuration was done in the account itself, so the VPC/subnets were "owned" by the AWS account itself. Now, the VPC owner is the central administration account. 

To deploy an ECS cluster into the VPC, we'd import the VPC like so: 
```
const vpc = awsx.ec2.Vpc.fromExistingIds('A VPC', {
  vpcId: VPC_ID,
  privateSubnetIds: SUBNET_IDS,
});
```

When trying to import a Shared VPC, we get the error:
```
    error: Preview failed: refreshing urn:pulumi:dev::pulumi-import-vpc::aws:ec2/vpc:Vpc::A VPC: 1 error occurred:
        * InvalidVpcID.NotFound: The vpc ID 'vpc-xxxxxx' does not exist
        status code: 400, request id: fb554de1-8352-4252-92e0-9883f74b13c4
```

Note that this is markedly different when passing in a junk VPC ID, like `vpc-foo`:
```
    error: Preview failed: resource 'vpc-foo' does not exist
```

This error seems to be stemming from the underlying provider: https://github.com/hashicorp/terraform-provider-aws/issues/12513, which was unfortunately was closed in favor of https://github.com/hashicorp/terraform-provider-aws/issues/9061, whose conversation does not seem to be focused on _importing shared resources_ but simply on being able to tag them. 

# Workarounds Attempted
- There doesn't seem to be another way to deploy an ECS cluster into a VPC without first importing the target VPC. 
- I tried importing the VPC using the non-crosswalk libraries, but those objects weren't compatible. 
- Haven't really opened up the conversation around not using RAM to manage shared resources in the company
- Using `pulumi import aws:ec2/vpc:Vpc A_VPC vpc-xxxxx` fails with the same TF provider error
- In the interest of moving off Crosswalk (not something I'd really like to do), I was looking at https://www.pulumi.com/docs/reference/pkg/aws/ec2/vpc/#look-up, but was confused by the "provider ID" arg. 

Is there anything from a purely-Pulumi perspective we could try that might work?

If not, it seems like until the appropriate fix is pushed through the underlying provider, Crosswalk won't be usable in RAM-managed accounts.

**Comment by binhrobles** [comment]
Also, I'm operating on the implicit assumption that Crosswalk uses terraform-provider-aws under the hood, but I'm not 100% on how that works.

**Comment by leezen** [comment]
Crosswalk (`awsx`) uses `@pulumi/aws` which uses `terraform-provider-aws`
In this particular case, an 'import' isn't actually an import, but a read, which ends up invoking the `get`:
https://github.com/pulumi/pulumi-awsx/blob/master/nodejs/awsx/ec2/vpc.ts#L77

In the case of a shared VPC, I don't think you actually want to import it. Importing a resource into a stack implies you're going to manage the state of that resource (including future modifications) for that resources. Instead, you just want to read the VPC to use it so that you can provide it to ECS.

A few thoughts:
1. Have you confirmed that the account credentials you're passing in pulumi have the right permissions to read that VPC?
1. If you want to use `@pulumi/aws` without Crosswalk, you should be able to pass in the VPC ID to ECS without ever having to do a `get` to have a VPC resource. You can just treat the VPC ID as a string.

**Comment by binhrobles** [comment]
Hey, thanks for the time.

So terminology-wise: `awsx.ec2.Vpc.fromExistingIds()` is a _read_ not an _import_. And so the issue is really that the underlying provider is having trouble reading/finding the shared VPC. Which makes sense that we'd want to debug from a permissions perspective first.

1. Yes, but do you have a specific operation in mind? With these creds I'm able to see the shared VPC w/ `aws ec2 describe-vpcs --vpc-id xxxxx`, and am able to use `fromExistingIds()` in the other account using the equivalent user. I guess the mismatch would be if there's a specific permission that needs to be enabled to read a shared VPC at some deeper level required by the underlying `aws.ec2.Vpc.get()` call, but I don't know enough about the way RAM-managed resources work or what operations the `Vpc.get()` call needs.

2. So if we moved off Crosswalk, we'd want to use https://www.pulumi.com/docs/reference/pkg/aws/ecs/service/#servicenetworkconfiguration and just drop our subnets right in there as strings and that could potentially work with the shared subnets.

**Comment by leezen** [comment]
> So terminology-wise: awsx.ec2.Vpc.fromExistingIds() is a read not an import. And so the issue is really that the underlying provider is having trouble reading/finding the shared VPC. Which makes sense that we'd want to debug from a permissions perspective first.

Correct

> I'm able to see the shared VPC w/ aws ec2 describe-vpcs --vpc-id xxxxx, and am able to use fromExistingIds() in the other account using the equivalent user.

That's pretty helpful actually. The underlying call is indeed DescribeVpcs against the AWS Go SDK (https://github.com/pulumi/terraform-provider-aws/blob/upstream-v3.17.0/aws/resource_aws_vpc.go#L574)

You can actually see the error you're getting comes from there as well a couple lines later.

So I'm surprised the `fromExistingIds` doesn't "just work" given. Unfortunately, we'd likely have to spend more time debugging and trying to reproduce at this point. If you're able to provide some debug logs, that might be helpful (redacting any sensitive information, of course). https://www.pulumi.com/docs/troubleshooting/#verbose-logging

I'm curious if you just invoke `const vpc = new aws.ec2.Vpc.get("foo", "vpc-id123", {})` if this works for you or if it fails in the same way? I would expect it to fail the same way, but just want to double-check.

> So if we moved off Crosswalk, we'd want to use https://www.pulumi.com/docs/reference/pkg/aws/ecs/service/#servicenetworkconfiguration and just drop our subnets right in there as strings and that could potentially work with the shared subnets.

Right

**Comment by binhrobles** [comment]
Yes, using `aws.ec2.Vpc.get()` snippet, seeing the same error: `InvalidVpcID.NotFound: The vpc ID 'vpc-xxxx' does not exist`.

I ran the `fromExistingIds()` code with TF and Pulumi trace logging and got some interesting results!
- The Go SDK successfully reads/returns a response for the DescribeSubnets AND DescribeVpcs requests, properly returning each of the 3 subnets and VPC attributes.
- It also runs through the call to `Action=DescribeVpcAttribute&Attribute=enableDnsSupport` successfully
- It then runs through some unmarshaling/marshaling of these resources. Nothing seems out of place.
- It emits a `ReadResource RPC finished: resource:Dev VPC-private-2[aws:ec2/subnet:Subnet]` for each of the subnets, with seemingly correct details.
- We finally get our error, with some more details:
```
  I1125 11:12:15.959592   42823 provider_plugin.go:1354] rpc error kind `Unknown` may not be recoverable
  I1125 11:12:15.959647   42823 provider_plugin.go:820] Provider[aws, 0xc0006a58c0].Read(vpc-03b7e8843df0ce017,urn:pulumi:dev::pulumi-import-vp  c::aws:ec2/vpc:Vpc::Dev VPC) failed: rpc error: code = Unknown desc = refreshing urn:pulumi:dev::pulumi-import-vpc::aws:ec2/vpc:Vpc::Dev VPC:   1 error occurred:
    * InvalidVpcID.NotFound: The vpc ID 'vpc-xxxxx' does not exist
    status code: 400, request id: 94b288dc-4d83-4b60-a7ef-72c29e86ab9f
```

So I'm not super sure what's happening. Seems like the AWS API calls themselves are working, but something happens after that that breaks and returns an rpc error `Unknown`.
